### PR TITLE
[d3-array]: Implement unlimited number of accessor key functions for group, groups, etc. functions

### PR DESCRIPTION
**Pull request by kmcoding**
* Add new exported types: `NestedInternMap` and `NestedArray`
* Call and type signatures for the following functions have been updated to allow a (theoretically) unlimited number of accessor key functions:
  - group
  - groups
  - flatGroup
  - index
  - indexes
  - rollup
  - rollups
  - flatRollup
* The parameter signature for accessor key functions now correctly takes in at most three parameters: the value, the current index and all values (this was previously missing and was mentioned in this discussion: https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/60917)
* Add test cases for more than three accessor key functions (due to the `max-line-length  Exceeds maximum line length of 200` restriction by dtslint, adding test cases with even more accessor functions is currently not possible)

Fixes the issues mentioned in the following discussion/issues:
- https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/62830
- https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/57101
- https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/57356
- https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/60917

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <https://d3js.org/d3-array/group>
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

**Comment by DangerBotOSS** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## d3-array ([<kbd>unpkg</kbd>](https://unpkg.com/browse/d3-array@latest/))
was missing the following properties:
1. scan
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 1a3a49001342b1de649ea8ee367a2dc135744dd7
</p>

**Review by Fil — COMMENTED** [review]
Can you make two commits: one for changes, the other for formatting? It's hard to review 1682 lines.

**Comment by kmcoding** [comment]
> Can you make two commits: one for changes, the other for formatting? It's hard to review 1682 lines.

Just pushed two separate commits. Most of the new lines are due to the extended test cases with 10 or more accessor key functions. I hope it is better to review now. Let me know if you need me to do any other changes!

File: types/d3-array/d3-array-tests.ts
Comment (jakebailey) [inline types/d3-array/d3-array-tests.ts:815]: Reviewing this would be a bit easier if this change didn't include a load of unrelated formatting changes; I do see what's happening in the index and it's probably fine, but the tests really have these looking huge and different.

**Review by jakebailey — COMMENTED** [review]
I'm not totally sure I'm qualified to review this, but given this is a critical package and you're "breaking" the API by changing the overloads / type parameters, is it definitely the case that nobody is referring to these with specific type parameters?

Maybe this new overload needs to be added at the bottom or something?

**Comment by sandersn** [comment]
Unfortunately, we can't merge a change to this critical package without an approval from one or two maintainers

@gustavderdrache, @borisyankov, @tomwanzek, @denisname, @Ledragon, @Methuselah96, @Fil can you help out? @Fil I know you looked at this earlier.

**Review by Methuselah96 — APPROVED** [review]
Thanks for working on this, sorry for the delayed review! Seems like there might be a more maintainable way to express the tests, but we can take that up in a future PR if necessary.

**Comment by Methuselah96** [comment]
I tested that that the new code works with the tests before this PR.

File: types/d3-array/index.d.ts
Comment (Methuselah96) [inline types/d3-array/index.d.ts:499]: I'm curious if it's possible to flip this on it's head and make it be `TKeys extends unknown[]` and then map the `TKeys` to functions in the parameter.

**Review by Methuselah96 — CHANGES_REQUESTED** [review]

**Comment by kmcoding** [comment]
> Thanks for working on this, sorry for the delayed review! Seems like there might be a more maintainable way to express the tests, but we can take that up in a future PR if necessary.

@Methuselah96 thanks for your review! I am with you on this, however there was, at least to my current knowledge, no other way in expressing the new tests other than I have implemented them.

If I will find any other way of writing the tests for cases with lots of key accessor functions, then I will make sure to open another PR for that.

File: types/d3-array/index.d.ts
Comment (kmcoding) [reply types/d3-array/index.d.ts:499]: > I'm curious if it's possible to flip this on it's head and make it be `TKeys extends unknown[]` and then map the `TKeys` to functions in the parameter. Seems like that would be a lot easier for people who need to manually specify the type parameters.

@Methuselah96 I will probably find some time on the weekend to take a look at your idea! I will report back to you then

Comment (Methuselah96) [reply types/d3-array/index.d.ts:499]: I took a quick stab at this and something like this seems promising:

```ts
export type NestedInternMap<
  TObject,
  TReduce,
  TKeys extends unknown[],
> = TKeys extends [infer TFirst, ...infer TRest]
  ? InternMap<TFirst, NestedInternMap<TObject, TReduce, TRest>>
  : TReduce;

export function group<TObject, TKeys extends unknown[]>(
  iterable: Iterable<TObject>,
  ...keys: {
    [Index in keyof TKeys]: (
      value: TObject,
      index: number,
      values: TObject[],
    ) => TKeys[Index];
  }
): NestedInternMap<TObject, TObject[], TKeys>;

```

Comment (kmcoding) [reply types/d3-array/index.d.ts]: @Methuselah96 I have committed new changes to this PR, taking your ideas as a basis for the refactoring. The way you expressed the types is way more efficient than my previous implementation and more readable as well 👍 I have also added new test cases for explicitly providing the type parameters to each function and tested those via the `// $ExpectType ...` syntax. Could you have a look at the changes again?

Thanks for your support on this!

**Review by Fil — APPROVED** [review]

**Review by Methuselah96 — APPROVED** [review]

**Review by jakebailey — APPROVED** [review]
If owners are good with it, then I'll approve to get this moving, but I still have no clue what's actually changing in the test code due to the mass of formatting changes, so my only option is to "trust and revert later if it breaks" 😢

**Comment by Methuselah96** [comment]
> I still have no clue what's actually changing in the test code due to the mass of formatting changes

I successfully ran the tests before all the formatting changes with the new code and everything passed so that gives me hope.

**Comment by pearmini** [comment]
Is this a breaking change but introduced by a patch version? The code snippet below works fine with `@types/d3-array@0.3.5`, but throws an error with `@types/d3-array@0.3.7`.

```ts
import { group } from 'd3-array';

const I = [0, 1, 2, 3, 4, 5];

// !!! Error: Type 'boolean' does not satisfy the constraint 'unknown[]'
const groups = group<number, boolean>(I, (i) => i % 2 === 0);
```

**Comment by Methuselah96** [comment]
Yeah, types on DefinitelyTyped don't generally follow semantic versioning because the versions match with the JS package and avoiding breaking changes in types is harder than avoiding breaking changes in JS.

In general, the way to fix the generics is to replace:

```ts
const groups = group<T1, T2>(...);
```

with:

```ts
const gropus = group<[T1, T2]>(...);
```

How many errors are you encountering with this upgrade? I'm noting that in the example you provided that I don't think you need the generics at all, and should be able to just do:

```ts
const groups = group(I, (i) => i % 2 === 0);
```

We do have the option of readding function overloads for the previous definition, but I'd want to see this causing consistent pain with more users before doing that. Or you can create a PR to do so if you feel strongly about it.

**Comment by pearmini** [comment]
@Methuselah96 Thanks for reply!

The errors introduced by this upgrade can be easily fixed in my project. And the example I provided is only for showing the breaking change, it is indeed given with redundant generics.

Currently my problem is: **the new generics is really better than the old one**? Surely the new generics provide the ability to specify unlimited number of accessor key, but I think it introduces more complexity because it is inconsistent with the JS API.

```js
// The actual way:
const groups = group(array, d => d.key1, d => d.key2, d => d.key3);

// Not this way:
const groups = group(array, [d => d.key1, d => d.key2, d => d.key3]);
```

It is especially confusing when there is only one key:

```ts
// Why need an array generics?
const groups = group<TObject, [Key]>(array, d => d.key1);
```

I' think the actual problem before is that the function overloads are not enough, not needing unlimited number of accessor key. So why just add more key parameters(maybe from 3 to 9)? Then makes the old generics the primary option. And uses the new generics only for extreme situation, e.g., the number of keys is more 9.

**Comment by Methuselah96** [comment]
I understand your point of view, but I think there are some other reasons to consider for sticking with the new generics.

1. 99% of the time the user shouldn't have to specify the generics. The only reason I can think of to specify the generics is if the user wants to widen the inferred type of one of the keys, which seems rare.
2. Having a single function with no overloads is a little easier to maintain. This PR dropped almost 200 lines of type definitions (mostly JSDoc comments). Having to maintain the JSDoc comments across all the overloads is a PITA (and would be more so if there were more overloads).
3. You mentioned that the new way is "inconsistent with the JS API." However that's looking at it the perspective of each parameter being separate and not part of a rest parameter. If I were to express the method in JS it would look something like:
```js
function group(values, ...keys) {
  return nest(values, identity, identity, keys);
}
```
and the keys parameter does end up being an array. So there's no way to make the generics "consistent" with the JS API since the parameter is either an array or separate parameters depending on your point of view. Typescript [introduced generic rest parameters in TypeScript 3.0](https://devblogs.microsoft.com/typescript/announcing-typescript-3-0/#tuples-and-parameters) and framed as a way to no longer have to specify multiple overloads. The only reason these types weren't written this way to begin with is because these types were added before TypeScript 3.0.

I understand where you're coming from, I just wanted to provide some more reasoning for why the new way seems better to me.

**Comment by pearmini** [comment]
Thanks for the detailed explanation! I know your concerns, but I have one last thought. 

Is there any chance the generics will be written using rest parameters? Because I find out the [unknown top type](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-0.html#new-unknown-top-type) used in this PR is also introduced in TypeScript 3.0. If we can use features introduced in TypeScript 3.0, why can't we use more?

**Comment by Methuselah96** [comment]
I'm not sure I understand your question. This PR makes it so that we are using generic rest parameters. [DefinitelyTyped's support window](https://github.com/DefinitelyTyped/DefinitelyTyped#support-window) means we can use any features that were introduced in TypeScript 4.4 or earlier (including TypeScript 3.0).

**Comment by pearmini** [comment]
@Methuselah96

Oh, I've made a mistake. I mistakenly thought TS support syntax like this:

```ts
export function group<TObject, ...TKeys extends unknown[]> {}
```

then can specify unlimited number of key accessor:

```ts
const groups = group<Array, TKey1, TKey2, TKey3, ...>();
```

I'm not very good at TS, sorry to bother you. 😣

**Comment by Methuselah96** [comment]
No worries, happy to discuss. That would be cool if TypeScript had syntax like that. 😎

**Pull request merged**
